src/runtime_src/core/pcie/common/dmatest.h

```diff
@@ -80,74 +89,73 @@ namespace xcldev {
+                });
+        }
+
+        int validate() const {
+            std::unique_ptr<char[]> bufCmp(new char[mSize]);
+            std::memset(bufCmp.get(), pattern, mSize);
+            int result = 0;
+            for (const std::pair<xclBufferHandle, xrt_core::aligned_ptr_type> &bo : mBOList) {
+                if (!std::memcmp(bo.second.get(), bufCmp.get(), mSize))
+                    continue;
+                throw xrt_core::error(-EIO, "DMA Test data integrity check failed.");
+            }
             return result;
         }
 
     public:
-        DMARunner(xclDeviceHandle handle, size_t size, unsigned flags=0) : mHandle(handle), mSize(size), mFlags(flags) {
+        DMARunner(xclDeviceHandle handle , size_t size, unsigned flags=0) : mHandle(handle), mSize(size),
+                                                                            mFlags(flags), pattern('x') {
             long long count = 0x100000000/size;
+
+            if (count == 0)
+                throw xrt_core::error(-EINVAL, "DMA buffer size cannot be larger than 0x100000000.");
+
             if (count > 0x40000)
                 count = 0x40000;
 
             for (long long i = 0; i < count; i++) {
-                auto bo = xclAllocBO(mHandle, mSize, 0, mFlags);
+                // This can throw and callers of DMARunner are supposed to catch this.
+                xrt_core::aligned_ptr_type buf = xrt_core::aligned_alloc(xrt_core::getpagesize(), mSize);
+                xclBufferHandle bo = xclAllocUserPtrBO(mHandle, buf.get(), mSize, mFlags);
                 if (bo == XRT_NULL_BO)
                     break;
-                mBOList.push_back(bo);
+                std::memset(buf.get(), pattern, mSize);
+                mBOList.emplace_back(bo, std::move(buf));
             }
+            if (mBOList.size() == 0)
+                throw xrt_core::error(-ENOMEM, "No DMA buffers could be allocated.");
         }
 
         ~DMARunner() {
-            for (auto bo : mBOList)
-                xclFreeBO(mHandle, bo);
-        }
-
-        int validate(const char *buf) const {
-            std::unique_ptr<char[]> bufCmp(new char[mSize]);
-            size_t result = 0;
-            for (auto i : mBOList) {
-                //Clear out the host buffer
-                std::memset(bufCmp.get(), 0, mSize);
-                result = xclReadBO(mHandle, i, bufCmp.get(), mSize, 0);
-                if (result) {
-                    std::cout << "DMA Test data integrity read failed with Error = " << result << "\n";
-                    break;
-		}
-
-                if (std::memcmp(buf, bufCmp.get(), mSize)) {
-                    std::cout << "DMA Test data integrity check failed\n";
-                    break;
-                }
-            }
-            return static_cast<int>(result);
+            std::for_each(mBOList.begin(), mBOList.end(), [&](std::pair<xclBufferHandle, xrt_core::aligned_ptr_type> &bo) {
```

Comment:
you know that lambda in C++14 takes auto as argument, e.g.
`std::for_each(mBOList.begin(), mBOList.end(), [&](auto& bo) { xclFreeBO(mHandle, bo.first); });`
This may fit on one line.

Reply:
Updated